src/rime/gear/single_char_filter.cc

```diff
@@ -55,11 +58,69 @@ bool SingleCharFirstTranslation::Rearrange() {
   return !cache_.empty();
 }
 
-SingleCharFilter::SingleCharFilter(const Ticket& ticket) : Filter(ticket) {}
+class SingleCharOnlyTranslation : public Translation {
+ public:
+  SingleCharOnlyTranslation(an<Translation> translation);
+
+  bool Next() override;
+  an<Candidate> Peek() override;
+
+ private:
+  an<Translation> translation_;
+  an<Candidate> current_;
+};
+
+SingleCharOnlyTranslation::SingleCharOnlyTranslation(
+    an<Translation> translation)
+    : translation_(translation) {}
```

Comment:
The library expects a `Translation` to be ready to query with `Peek()` or `exhausted()` in its initial state.
Not only in `Next()`, but also in the constructor, we need to locate the next available single-char candidate and set `exhausted_` depending on whether there is any single-char candidate available. Usually I extract a function from `Next` and call it in both places.

The interface will be improved in r2ime. Migrating the class to C++ iterator seems to be a large change.

Reply:
should be fixed. PTAL